### ISSUE TITLE
SolverProblemError after upgrade to poetry 1.1.5+

### ISSUE THREAD

**Issue by RafalSkolasinski** [python-poetry/poetry#4301]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Pop!_OS 21.04
- **Poetry version**: 1.1.4, 1.1.5, 1.1.7
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: [my pyproject.toml](https://gist.github.com/RafalSkolasinski/30892b156e11ebfb5aa22e3b8e855f75)

## Issue

I have an environment definition in form of `pyproject.toml, see gist above, that used to work perfectly fine with Poetry 1.1.4 (and probably some versions before).

When I upgraded to latest it started to fail, see bellow. I tested that it already starts to fail on 1.1.5.

<details>
<summary> When it works on poetry 1.1.4 </summary>

```
poetry install
Installing dependencies from lock file

Package operations: 118 installs, 0 updates, 0 removals

  • Installing ipython-genutils (0.2.0)
  • Installing attrs (20.3.0)
  • Installing pyrsistent (0.17.3)
  • Installing six (1.15.0)
  • Installing traitlets (5.0.5)
  • Installing jsonschema (3.2.0)
  • Installing jupyter-core (4.7.1)
  • Installing parso (0.7.1)
  • Installing ptyprocess (0.7.0)
  • Installing pyparsing (2.4.7)
  • Installing python-dateutil (2.8.1)
  • Installing pyzmq (22.0.2)
  • Installing tornado (6.1)
  • Installing wcwidth (0.2.5)
  • Installing async-generator (1.10)
  • Installing backcall (0.2.0)
  • Installing decorator (4.4.2)
  • Installing jedi (0.17.2)
  • Installing jupyter-client (6.1.11)
  • Installing markupsafe (1.1.1)
  • Installing nbformat (5.1.2)
  • Installing nest-asyncio (1.5.1)
  • Installing packaging (20.9)
  • Installing pexpect (4.8.0)
  • Installing pickleshare (0.7.5)
  • Installing prompt-toolkit (3.0.14)
  • Installing pycparser (2.20)
  • Installing pygments (2.7.4)
  • Installing webencodings (0.5.1)
  • Installing bleach (3.3.0)
  • Installing cffi (1.14.4)
  • Installing defusedxml (0.6.0)
  • Installing entrypoints (0.3)
  • Installing ipython (7.20.0)
  • Installing jinja2 (2.11.3)
  • Installing jupyterlab-pygments (0.1.2)
  • Installing mistune (0.8.4)
  • Installing nbclient (0.5.2)
  • Installing pandocfilters (1.4.3)
  • Installing testpath (0.4.4)
  • Installing argon2-cffi (20.1.0)
  • Installing certifi (2020.12.5)
  • Installing chardet (4.0.0)
  • Installing idna (2.10)
  • Installing ipykernel (5.4.3)
  • Installing nbconvert (6.0.7)
  • Installing numpy (1.20.1)
  • Installing prometheus-client (0.9.0)
  • Installing pytz (2021.1)
  • Installing send2trash (1.5.0)
  • Installing smmap (3.0.5)
  • Installing terminado (0.9.2)
  • Installing urllib3 (1.26.3)
  • Installing cycler (0.10.0)
  • Installing gitdb (4.0.5)
  • Installing h5py (2.10.0)
  • Installing iniconfig (1.1.1)
  • Installing json5 (0.9.5)
  • Installing kiwisolver (1.3.1)
  • Installing notebook (6.2.0)
  • Installing numexpr (2.7.2)
  • Installing pandas (1.2.2)
  • Installing pillow (8.1.0)
  • Installing pluggy (0.13.1)
  • Installing py (1.10.0)
  • Installing pyyaml (5.4.1)
  • Installing requests (2.25.1)
  • Installing sortedcontainers (2.3.0)
  • Installing toml (0.10.2)
  • Installing appdirs (1.4.4)
  • Installing atomicwrites (1.4.0)
  • Installing click (7.1.2)
  • Installing colorama (0.4.4)
  • Installing coverage (5.4)
  • Installing dask (2021.2.0)
  • Installing gitpython (3.1.13)
  • Installing h5netcdf (0.9.0)
  • Installing imageio (2.9.0)
  • Installing joblib (1.0.1)
  • Installing jupyterlab-server (1.2.0)
  • Installing jupyterlab-widgets (1.0.0)
  • Installing llvmlite (0.35.0rc3)
  • Installing matplotlib (3.3.4)
  • Installing mypy-extensions (0.4.3)
  • Installing networkx (2.5)
  • Installing pathspec (0.8.1)
  • Installing pyflakes (2.2.0)
  • Installing pywavelets (1.1.1)
  • Installing pytest (6.2.2)
  • Installing regex (2020.11.13)
  • Installing retrying (1.3.3)
  • Installing scipy (1.6.0)
  • Installing sortedcollections (2.1.0)
  • Installing tables (3.6.1)
  • Installing threadpoolctl (2.1.0)
  • Installing tifffile (2021.2.1)
  • Installing tqdm (4.56.1)
  • Installing typed-ast (1.4.2)
  • Installing typing-extensions (3.7.4.3)
  • Installing widgetsnbextension (3.5.1)
  • Installing xarray (0.16.2)
  • Installing adaptive (0.11.1)
  • Installing black (20.8b1)
  • Installing deepdish (0.3.6)
  • Installing ipywidgets (7.6.3)
  • Installing jupyterlab (2.2.9)
  • Installing line-profiler (3.1.0)
  • Installing nbdime (2.1.0)
  • Installing nbqa (0.5.7)
  • Installing numba (0.51.2)
  • Installing plotly (4.14.3)
  • Installing pytest-cov (2.11.1)
  • Installing pytest-flakes (4.0.3)
  • Installing scikit-image (0.17.2)
  • Installing scikit-learn (0.23.2)
  • Installing seaborn (0.11.1)
  • Installing sqlalchemy (1.3.23)
  • Installing xyzpy (1.0.0)

Installing the current project: ml-env (0.1.0)
```
</details>

<details>
<summary> When it fails on 1.1.5 </summary>

```
poetry install -vvv
Using virtualenv: /home/rskolasinski/.cache/pypoetry/virtualenvs/ml-env-Fs2Mjdk4-py3.8
Installing dependencies from lock file

Finding the necessary packages for the current system

  Stack trace:

  8  ~/.poetry/lib/poetry/_vendor/py3.8/clikit/console_application.py:131 in run
      129│             parsed_args = resolved_command.args
      130│ 
    → 131│             status_code = command.handle(parsed_args, io)
      132│         except KeyboardInterrupt:
      133│             status_code = 1

  7  ~/.poetry/lib/poetry/_vendor/py3.8/clikit/api/command/command.py:120 in handle
      118│     def handle(self, args, io):  # type: (Args, IO) -> int
      119│         try:
    → 120│             status_code = self._do_handle(args, io)
      121│         except KeyboardInterrupt:
      122│             if io.is_debug():

  6  ~/.poetry/lib/poetry/_vendor/py3.8/clikit/api/command/command.py:171 in _do_handle
      169│         handler_method = self._config.handler_method
      170│ 
    → 171│         return getattr(handler, handler_method)(args, io, self)
      172│ 
      173│     def __repr__(self):  # type: () -> str

  5  ~/.poetry/lib/poetry/_vendor/py3.8/cleo/commands/command.py:92 in wrap_handle
       90│         self._command = command
       91│ 
    →  92│         return self.handle()
       93│ 
       94│     def handle(self):  # type: () -> Optional[int]

  4  ~/.poetry/lib/poetry/console/commands/install.py:71 in handle
       69│         self._installer.verbose(self._io.is_verbose())
       70│ 
    →  71│         return_code = self._installer.run()
       72│ 
       73│         if return_code != 0:

  3  ~/.poetry/lib/poetry/installation/installer.py:103 in run
      101│         local_repo = Repository()
      102│ 
    → 103│         return self._do_install(local_repo)
      104│ 
      105│     def dry_run(self, dry_run=True):  # type: (bool) -> Installer

  2  ~/.poetry/lib/poetry/installation/installer.py:308 in _do_install
      306│ 
      307│         with solver.use_environment(self._env):
    → 308│             ops = solver.solve(use_latest=whitelist)
      309│ 
      310│         # We need to filter operations so that packages

  1  ~/.poetry/lib/poetry/puzzle/solver.py:65 in solve
       63│         with self._provider.progress():
       64│             start = time.time()
    →  65│             packages, depths = self._solve(use_latest=use_latest)
       66│             end = time.time()
       67│ 

  SolverProblemError

  Because no versions of numba match >0.51.2,<0.52.0
   and numba (0.51.2) depends on llvmlite (>=0.34.0.dev0,<0.35), numba (>=0.51.2,<0.52.0) requires llvmlite (>=0.34.0.dev0,<0.35).
  So, because no versions of llvmlite match >=0.34.0.dev0,<0.35
   and ml-env depends on numba (^0.51.2), version solving failed.

  at ~/.poetry/lib/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes
```
</details>

The error of latest poetry (today 1.1.7)
```
poetry install -vvv
Creating virtualenv ml-env-Fs2Mjdk4-py3.8 in /home/rskolasinski/.cache/pypoetry/virtualenvs
Using virtualenv: /home/rskolasinski/.cache/pypoetry/virtualenvs/ml-env-Fs2Mjdk4-py3.8
Installing dependencies from lock file

Finding the necessary packages for the current system

  Stack trace:

  8  ~/.poetry/lib/poetry/_vendor/py3.8/clikit/console_application.py:131 in run
      129│             parsed_args = resolved_command.args
      130│ 
    → 131│             status_code = command.handle(parsed_args, io)
      132│         except KeyboardInterrupt:
      133│             status_code = 1

  7  ~/.poetry/lib/poetry/_vendor/py3.8/clikit/api/command/command.py:120 in handle
      118│     def handle(self, args, io):  # type: (Args, IO) -> int
      119│         try:
    → 120│             status_code = self._do_handle(args, io)
      121│         except KeyboardInterrupt:
      122│             if io.is_debug():

  6  ~/.poetry/lib/poetry/_vendor/py3.8/clikit/api/command/command.py:171 in _do_handle
      169│         handler_method = self._config.handler_method
      170│ 
    → 171│         return getattr(handler, handler_method)(args, io, self)
      172│ 
      173│     def __repr__(self):  # type: () -> str

  5  ~/.poetry/lib/poetry/_vendor/py3.8/cleo/commands/command.py:92 in wrap_handle
       90│         self._command = command
       91│ 
    →  92│         return self.handle()
       93│ 
       94│     def handle(self):  # type: () -> Optional[int]

  4  ~/.poetry/lib/poetry/console/commands/install.py:71 in handle
       69│         self._installer.verbose(self._io.is_verbose())
       70│ 
    →  71│         return_code = self._installer.run()
       72│ 
       73│         if return_code != 0:

  3  ~/.poetry/lib/poetry/installation/installer.py:103 in run
      101│         local_repo = Repository()
      102│ 
    → 103│         return self._do_install(local_repo)
      104│ 
      105│     def dry_run(self, dry_run=True):  # type: (bool) -> Installer

  2  ~/.poetry/lib/poetry/installation/installer.py:305 in _do_install
      303│ 
      304│         with solver.use_environment(self._env):
    → 305│             ops = solver.solve(use_latest=self._whitelist)
      306│ 
      307│         # We need to filter operations so that packages

  1  ~/.poetry/lib/poetry/puzzle/solver.py:65 in solve
       63│         with self._provider.progress():
       64│             start = time.time()
    →  65│             packages, depths = self._solve(use_latest=use_latest)
       66│             end = time.time()
       67│ 

  SolverProblemError

  Because no versions of numba match >0.51.2,<0.52.0
   and numba (0.51.2) depends on llvmlite (>=0.34.0.dev0,<0.35), numba (>=0.51.2,<0.52.0) requires llvmlite (>=0.34.0.dev0,<0.35).
  So, because no versions of llvmlite match >=0.34.0.dev0,<0.35
   and ml-env depends on numba (^0.51.2), version solving failed.

  at ~/.poetry/lib/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes
```
```

**Comment by RafalSkolasinski** [comment]
Note, I also have a [poetry.lock](https://gist.github.com/RafalSkolasinski/0c72b98811424ef0d30d6eed97a6a661) file just next to `pyproject.toml` and would expect that if this was tested configuration that worked later version of poetry would just install what was there.

**Comment by finswimmer** [comment]
Hello @RafalSkolasinski,

the problem is your lock file. Please regenerate it with `poetry lock --no-update` or `poetry lock` and try again.

fin swimmer

**Comment by RafalSkolasinski** [comment]
Thanks @finswimmer, this actually solves. The only meaningful difference in deps I can spot now is usage of 
`0.34.0` version for `llvmlite`. 

I suspect that reason is because earlier version of poetry wrongly included `0.35.0rc3` of `llvmlite` as `<0.35`?

**Comment by finswimmer** [comment]
> I suspect that reason is because earlier version of poetry wrongly included 0.35.0rc3 of llvmlite as <0.35?

Could be :man_shrugging: :)
